### DISCUSSION TITLE
toggling in pocketbase

### DISCUSSION THREAD

**Discussion by thal-sh** [pocketbase/pocketbase#2956] [Q&A] [answered]
i am making a social media post, the like button should be toggled, my question is how can i implement that? i tried some ways but they made the process really slow, (keep in mind that when the user comes back he'll be able to see if he liked the post or not) so it doesnt reset. I tried doing that that by creating a relationship between posts table and likes table, but each time the user likes i have to check if a record already exists or not, by filtering, which is slow.. any help?

**Comment by ganigeorgiev** [comment]
>  but each time the user likes i have to check if a record already exists or not

How did you perform the check? In general you don't have to do any manual checks. If you have a "likes" table that contains 2 single `relation` fields (eg. "post" and "user"), then you can create a combined unique index consisting of those 2 fields. The unique index can be created from the collection panel, right under the fields:

![index_screenshot](https://i.imgur.com/nNpKR25.png)

**Reply by thal-sh** [reply]
```javascript
async function handleLike() {
    try {
      const auth = pb.authStore.model?.id;
      if (auth) {
        const resultList = await pb.collection("likes").getList(1, 1, {
          filter: `user_id = "${pb.authStore.model?.id}" && post_id = "${post.id}"`,
        });
        console.log(resultList.items[0]?.id);
        if (resultList.totalItems === 0) {
          try {
            const likeRecord = {
              user_id: '3lzcgv2imfe7eba',
              post_id: post.id,
              liked: true,
            };
            const like = await pb.collection("likes").create(likeRecord);
            const data = {
              likes: likes + 1,
            };
            const record = await pb.collection("posts").update(post.id, data);
```
if the returned value is 0, then there is no record in this array.
can u explain what tables should i create, what i did is have a post table and like table, both have a relationship together
to check for likes i search in the likes table for user_id and post_id...if it doesnt exist, add a new record and increment the llikes in posts, if not then decrement the likes from the post table and delete the record

**Reply by thal-sh** [reply]
also one more thing that you'd understand the issue, my issue isnt making a relation, my issue is toggling, for each post a user can like, he can increment the like btn once, if he clicks it again it should decrement, im finding it hard to make this logic, i tried making it that why (in the previous code) but it really slow and kinda not good, and i cant know if a user liked a certain post or not without checking if the record is in the likes table or not, i have to check so i'd decide to either increment or decrement

**Reply by ganigeorgiev** [reply, marked as answer]
To avoid the manual cheks in your `handleLike` function, you can create a **unique combined index** with the `user_id` and `post_id` fields. This way when a user makes an attempt to like a post second time it will throw an error.

Additionally, I'd recommend not to manually count the likes because the data can get easily out of sync this way. Instead, you can create a helper [view collection](https://pocketbase.io/docs/collections/#view-collection) that will perform the count for you. The view query could look something like:
```sql
-- eg. the collection could be named something like "totalLikes"
select
    post_id as id,
    post_id, -- optional, adds post_id as relation field
    count(id) as total
from likes
group by (post_id)
```
The above will produce a new read-only collection with 2 fields `post_id` and `total`.

To determine whether the user already liked a post and show some indicator in your UI, there are several ways you can do this, but one possible and efficient approach I immagine could be to eager fetch all currently loaded posts and their post-user state upfront (aka. in total you'll have only 3 requests):

```js
// fetch the posts that you want to show
const posts = await pb.collection("posts").getList(1, 30);

// construct a filter expression from the fetched post ids `id="id1" || id="id2" ...`
const postsFilter = posts.map((p) => `id="${p.id}"`).join("||");

// fetch all likes for the current user and the previously fetched posts
const currentUserLikes = await pb.collection("likes").getFullList({
    filter: `user_id="${pb.authStore.model?.id}" && (${postsFilter})`
})

// fetch a list with all posts total likes
const totalLikes = await pb.collection("totalLikes").getFullList({
    filter: postsFilter,
})

....


// list the posts with their current user like status and total likes count
for (post of posts) {
    const total = totalLikes.find((r) => r.post_id == post.id)?.total || 0;
    const isLiked = !!currentUserLikes.find((r) => r.post_id == post.id);

    console.log("Post ID", post.id)
    console.log("Total likes", total)
    console.log("Is liked by the current user", isLiked)
}
```

When the user click on a like for a single post, you can trigger a handler like:

```js
async function likeToggle1(post) {
    const like = currentUserLikes.find((r) => r.post_id == post.id);
    if (like) {
        // send a "unlike" request (aka. delete)
        await pb.collection("likes").delete(like.id)
    } else {
        // send a "like" request (aka. create)
        await pb.collection("likes").create({
            post_id: post.id,
            user_id: pb.authStore.model?.id,
        })
    }

    // trigger totalLikes and currentUserLikes reload...
}

// or alternatively, if you don't want to check the currentUserLikes array,
// you can rely on the unique index error to determine whether to like/unlike a post
// (note that this may not be as efficient as the previous version)
async function likeToggle2(post) {
    try {
        // try to like the post
        // (it will fail if it is already liked)
        await pb.collection("likes").create({
            post_id: post.id,
            user_id: pb.authStore.model?.id,
        })
    } catch (_) {
        // the post was already liked -> find and delete the like
        const like = await pb.collection("likes").getFirstListItem(`post_id="${post.id}" && user_id="${pb.authStore.model?.id}"`)

        await pb.collection("likes").delete(like.id)
    }

    // trigger totalLikes and currentUserLikes reload...
}
```

**Reply by thal-sh** [reply]
that's awesome, ill try adding these to my code, and get back to u

**Reply by ganigeorgiev** [reply]
>  but ive seen some stuff like .find() that isnt in js, is it go?

No, `find()` is a JS array method - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find (_it is just an example, you can use any other retrieval method you want_).

**Reply by thal-sh** [reply]
can't we do if the user tries to like again and he gets an error, can't we delete the record? so like we let the user create sth but if it is already there he'll get an error cuz it is unique so in that case i'll delete the record, if it is possible how can i do it?

**Reply by ganigeorgiev** [reply]
@karim-maw I think this is what the above `likeToggle2()` example do.

_Edit: I'm missing the code where the retrieval is done, I'll update it shortly_

**Reply by thal-sh** [reply]
thanks gani ill try to use these and so how it goes, if all good ill mark it as answer, thanks a lot, i love pocketbase and the support

**Reply by thal-sh** [reply]
okay so things work fine, tho one thing for future ppl who might see this, this would give u an error
```javascript
 const like = await pb.collection("likes").getFirstListItem({
            filter: `post_id="${post.id}" && user_id="${pb.authStore.model?.id}"`
        })
```
cuz it is expecting a filter of type string, that's how the function works, i did instead this and it worked 
```javascript
const like = await pb.collection("likes").getList(1, 1, {
        filter: `user_id = "${pb.authStore.model?.id}" && post_id = "${post.id}"`,
      });
// and then you delete with this
      await pb.collection("likes").delete(like.items[0].id)

```

**Reply by ganigeorgiev** [reply]
@karim-maw Oops, sorry I've updated the example; [`getFirstListItem(filter)`](https://github.com/pocketbase/js-sdk#crud-handlers) accept directly the filter as string, aka.:

```js
const like = await pb.collection("likes").getFirstListItem(`post_id="${post.id}" && user_id="${pb.authStore.model?.id}"`)
```
`getList(1, 1, { filter: ...})` should work also (the above is just a helper that internally calls `getList`).
